unliftio/src/UnliftIO/Async.hs

```diff
@@ -342,9 +339,536 @@ instance (MonadUnliftIO m, Semigroup a) => Semigroup (Concurrently m a) where
+        pure $ FlatApp $ FlatLift f a
+      both (LiftA2 f ca cb) = do
+        a <- both ca
+        b <- both cb
+        pure $ FlatApp $ FlatLiftA2 f a b
+      both (Alt ca cb) = do
+        a <- alt ca
+        b <- alt cb
+        case dlistToList (a `dlistConcat` b) of
+          []    -> E.throwIO EmptyWithNoAlternative
+          [x]   -> pure $ FlatApp x
+          x:y:z -> pure $ FlatAlt x y z
+      both (Pure a) = pure $ FlatApp $ FlatPure a
+
+      -- Returns a difference list for cheaper concatenation
+      alt :: forall k. Conc m k -> IO (DList (FlatApp k))
+      alt Empty = pure dlistEmpty
+      alt (Lift cf ca) = do
+        f <- both cf
+        a <- both ca
+        pure (dlistSingleton $ FlatLift f a)
+      alt (Alt ca cb) = do
+        a <- alt ca
+        b <- alt cb
+        pure $ a `dlistConcat` b
+      alt (Action m) = pure (dlistSingleton $ FlatAction (run m))
+      alt (LiftA2 f ca cb) = do
+        a <- both ca
+        b <- both cb
+        pure (dlistSingleton $ FlatLiftA2 f a b)
+      alt (Pure a) = pure (dlistSingleton $ FlatPure a)
+
+  both c0
+
+-- TODO: Write test that verifies that FlatAlt is actually Flat
+-- ((conc a <|> conc b) <|> conc c) <|> (conc d <|> conc e)
+
+-- | Run a @Flat a@ on multiple threads.
+runFlat :: Flat a -> IO a
+
+-- Silly, simple optimizations
+runFlat (FlatApp (FlatAction io)) = io
+runFlat (FlatApp (FlatPure x)) = pure x
+
+-- Start off with all exceptions masked so we can install proper cleanup.
+runFlat f0 = E.uninterruptibleMask $ \restore -> do
+  -- How many threads have started? We need to ensure we kill all
+  -- child threads and wait for them to die.
+  resultCountVar <- newTVarIO 0
+
+  -- Forks off as many threads as necessary to run the given Flat a,
+  -- and returns:
+  --
+  -- + An STM action that will block until completion and return the
+  --   result.
+  --
+  -- + The IDs of all forked threads. These need to be tracked so they
+  --   can be killed (either when an exception is thrown, or when one
+  --   of the alternatives completes first).
+  --
+  -- It would be nice to have the returned STM action return an Either
+  -- and keep the SomeException values somewhat explicit, but in all
+  -- my testing this absolutely kills performance. Instead, we're
+  -- going to use a hack of providing a TMVar to fill up with a
+  -- SomeException when things fail.
+  --
+  -- TODO: Investigate why performance degradation on Either
+  let go :: forall a.
+            TMVar E.SomeException
+         -> Flat a
+         -> IO (STM a, DList C.ThreadId)
+      go _excVar (FlatApp (FlatPure x)) = pure (pure x, dlistEmpty)
+      go excVar (FlatApp (FlatAction io)) = do
+        resVar <- newEmptyTMVarIO
+        tid <- C.forkIOWithUnmask $ \restore1 -> do
+          res <- E.try $ restore1 io
+          atomically $ do
+            -- TODO: What is the performance of this modifyTVar on high
+            -- contention?
+            modifyTVar' resultCountVar (+ 1)
```

Comment:
@snoyberg given this variable is going to cause more contention than any other variable in the block, what are your thoughts of moving this to another `atomically` call that happens after this one. I can't see any harm on not retrying the `tryPutTMVar` and `putTMVar` operations here, but I may be missing something.

Reply:
To make sure I understand the claim here: having an accurate `resultCountVar` is only an optimization, because if (for example) an exception somehow prevented this `+ 1` from running, it would only result in the thread kill code getting run when it isn't necessary. Is that the idea?

Reply:
Let me clarify, this is not what I'm trying to convey.

**TL;DR** - Discard this comment

---

I was not expecting the `resultCountVar` to have an inaccurate value given I was assuming there is no possibility of an async exception happening on this group of calls because of the `MaskedUninterruptible` state inside the sub-routine used in the `C.forkIOWithUnmask`. I went ahead and tested this assumption on a minimal example:

```haskell
module Main where

-- import RIO
import Control.Exception
import Control.Concurrent.STM
import Control.Concurrent.MVar
import Control.Concurrent

main :: IO ()
main = do
  lock <- newEmptyMVar
  accVar1 <- newTVarIO (0 :: Int)
  accVar2 <- newTVarIO (0 :: Int)
  tid <-
    uninterruptibleMask_ $
    forkIOWithUnmask $ \restore -> do
      putStrLn "1"
      getMaskingState >>= print
      putStrLn "2"
      atomically $ modifyTVar' accVar1 (+1)
      putStrLn "3"
      takeMVar lock
      putStrLn "4"
      atomically $ modifyTVar' accVar2 (+1)
      putStrLn "5"
      putMVar lock ()

  threadDelay 100
  killThread tid
  -- throwTo tid (ErrorCall "aha")
  putMVar lock ()
  takeMVar lock
  (atomically $ (,) <$> readTVar accVar1 <*> readTVar accVar2) >>= print
```

(compiled with `-threaded`)

I was expecting despite the fact that I do a `killThread` or `throwTo` to the spawned thread, that the sub-routine will continue running (printing 4, 5, and then increasing the count of `accVar2`) because I'm not calling `restore`, but, it seems I do not understand masking as well as I thought.

The actual output on the program gives returns is:

```
1
MaskedUninterruptible
2
3
(1,0)
```

Thanks for pointing me in the right direction.

Reply:
Noticing if I have the `atomically` call for `accVar2` on a `finally` block, I get the behavior I'm expecting, so this could be a possibility as well.